### ISSUE TITLE
Improve terminal UI

### ISSUE THREAD

**Issue by JeroenVinke** [monterey-framework/monterey#312]
https://github.com/monterey-framework/monterey/pull/304#issuecomment-247560595

Will add screen designs to this issue

**Comment by vegarringdal** [comment]
:+1:

**Comment by JeroenVinke** [comment]
When you click on the terminal icon, it will create a terminal immediately if there is no other existing terminal:

![image](https://cloud.githubusercontent.com/assets/2189477/18606692/505ee022-7cb7-11e6-916b-536d291742cf.png)

Then when you hide the terminal, and want to create another one by clicking on the terminal icon, you get this small popup:

![image](https://cloud.githubusercontent.com/assets/2189477/18606686/085db802-7cb7-11e6-9431-5f50b6c2e3a6.png)

would this be an improvement?

**Comment by Thanood** [comment]
I like it.
Could the project list (second screenshot) be a dropdown (dropup in that case)? That would mean fewer usage "interruptions" (you don't get a modal which changes background color and you don't need to move the pointer to the middle of the screen).

**Comment by JeroenVinke** [comment]
> Could the terminal list (second screenshot) be a dropdown (dropup in that case)? That would mean fewer usage "interruptions". With that I mean you don't get a modal which changes background color and visually fills the whole screen. And you don't need to move the pointer to the middle of the screen.

That would be nice, yeah

![image](https://cloud.githubusercontent.com/assets/2189477/18607931/77c6d89e-7cdb-11e6-8901-ac261b67e04c.png)

> "New Terminal" would look like a terminal row with a slightly different look (f.i. background color or font color, maybe border).

Can you expand on this a bit?

**Comment by Thanood** [comment]
![image](https://cloud.githubusercontent.com/assets/677826/18608140/27588ed8-7ce1-11e6-88f8-1009df0edaf9.png)

Instead of the button. Matter of taste but I like that better. 😃

**Comment by JeroenVinke** [comment]
Me too, it makes sense 👍

**Comment by adriatic** [comment]
... and me as well.

**Comment by adriatic** [comment]
Sorry for missing this issue - as I am likely to be opinionated once I fully understand what is being proposed. To help asking questions at this time (when my text is spatially far from the original) I took the liberty to edit @JeroenVinke's text by adding image labels.

Question 1: is the Terminal on Image 1 modal?
Question 2: is the Terminal on Image 1 in Monterey's child window?

Either yes answer makes me unhappy, as I do not like overlapping windows that do not allow me to at least move them away to uncover what is underneath. I might be too used to Visual Studio that tris really hard to allow you to see all windows that are **directly** involved with something I am currently doing. 

Question 3. Is the Terminal on image 1 a "real" terminal or just a blank panel (as it someone explained to me before)? If its not an embedded instance of Cmder (better to say: if it could not be an instance of Cmder) I do not find it useful.

I will likely have more comments once I understand the purpose of this feature - as I do not yet.

**Comment by vegarringdal** [comment]
1/2: yes, its a big dialog/modal on the screen (the one we have atm) sorry, the suggestion @JeroenVinke / @Thanood  have suggested looks like a very good solution to make it better.


3: 
Yes and no.
Its a special child process kinda like nodes child process only it has the settings that makes everthing we run it it to think its a real terminal , its using the pty.js (native module) that almost made @JeroenVinke  go crays and dump the entire module 😂 
Its also using a module called xterm.js for the display/colouring typing etc.

Both are the same visual studio code use. (maybe they have updated versions)

I havent checked the code lately been busy with another project, but the pty.js terminal and xtermjs could prb be used elsewhere is we are using nodes child process for any of the npm/jspm operations to make the task manager display what a normal terminal would have done

Hope some of this helped :-)

Links:

[xterm.js](https://github.com/sourcelair/xterm.js)
[pty.js](https://github.com/Tyriar/pty.js)


Last time I checked if a gulp watch etc was lunched with monterey & pty.js, and monterey failed/forced quit, pty.js also quit, so no weird background work after, dunno if we had a issue with this earlier

**Comment by JeroenVinke** [comment]
> Question 1: is the Terminal on Image 1 modal?

We can (and should) make this modeless

Question 2: is the Terminal on Image 1 in Monterey's child window?
Not sure what you mean by Monterey's child window. It will open a terminal in a dialog, not in a new window like it does when you click on the CMD.exe/CMDer tile

> Question 3. Is the Terminal on image 1 a "real" terminal or just a blank panel 
It is a real terminal. On Mac you get the Mac Terminal, and on Windows you get the real CMD.exe

> If its not an embedded instance of Cmder (better to say: if it could not be an instance of Cmder) I do not find it useful.
> I will likely have more comments once I understand the purpose of this feature - as I do not yet.

We can already start new terminals in Monterey:
![image](https://cloud.githubusercontent.com/assets/2189477/19019208/6e9f3c70-887f-11e6-930b-116f95c30b35.png)

But I hardly use it because it takes too many clicks in order to execute a command. If you want to execute a command **quickly**, and you don't want to open cmd.exe/cmder in a new window by clicking one of the tiles, then you can use the terminal **inside** Monterey. But right now it's too much "work" to start this terminal, execute a command en then close it again.

**Comment by vegarringdal** [comment]
Maybe a option to have own window for this could be useful.
Atm Im not using Monterey for running gulp watch since its so easy to just open prompt and write ```gulp watch```

Btw, if you click on the header on the running terminal you can also edit the name, so if you have several terminals you can give each one a name

I use visual studio code when coding, it also have a terminal thats very easy to run in current project, but Ive never used it for running gulp watch, but that would be simpler for me to use then open Monterey , opening task manager, then clicking gulp watch
Could be that I forget visual studio got the terminal that been keeping me from using it 😂 

**Think features for maintaining projects/installing plugins/installing libs/other skeletons like plugin-skeleton/bridge will be features that users will enjoy/notice more then a better terminal in the beginning.**

**Comment by JeroenVinke** [comment]
So the other option is to remove the terminal functionality and always have users start command prompts via the tiles (cmd.exe tile or cmder tile)

**Comment by vegarringdal** [comment]
Yes.
Maybe the button where you open the terminal now should just open the default prompt for the OS, so you wont have to leave current screen.
But might be better to ask users when Monterey have more users,  what they want/dont use etc.
Just wont not focus any time on the terminal atm over features that help you maintain the project.

**Comment by JeroenVinke** [comment]
Yep. We can keep this open in the meantime

**Comment by adriatic** [comment]
I agree with Vegar - at least at the moment (when I do not see the future workflows we might add) the comparison between "internal" and "external" terminal speaks in favor of external:

- Internal terminal starts in the proper folder.
- External terminal does not cover any Monterey windows.

In this situation I will **always** pick the external terminal despite the fact that it bugs me to have to set it's current folder.

I will add this discussion to the survey (note that we will change / enhance the current survey as the time goes buy and we learn more from the users)

**Comment by vegarringdal** [comment]
> In this situation I will always pick the external terminal despite the fact that it bugs me to have to set it's current folder.

Hold shift down while right clicking on the folder of project, and you can open the terminal with that path (just in case you didnt know) 

![image](https://cloud.githubusercontent.com/assets/2901416/19020595/e0366718-88ad-11e6-81bd-2110df78182a.png)

**Comment by adriatic** [comment]
Yes, I know this - and as you realize, this require having files explorer opened :-)

**Comment by vegarringdal** [comment]
yes :-)
Ive been testing the terminal in visual code, atm for gulp watch this was kinda nice, since it follows the project Im coding in
Maybe this would be practical for you too.
